### PR TITLE
Fix MaxPool implementation

### PR DESCRIPTION
**Pull request by tvlenin**
## Motivation and Context
MaxPool implementation had some bugs

## Description

- Change the interface in `dlk/python/dlk/core/view.py` to add compatibility with the MaxPool_struct
- Update the MaxPool parameters to add `T_UINT input_depth` necessary to do the MaxPool
-Modify the MaxPool Algorithm in ` dlk/python/dlk/templates/src/func/max_pool.cpp`

## How has this been tested?

Was tested with lmnet_v1.py using maxpool instead of averagePool
Tested on x86, ARM, FPGA. All test successful 
## Types of changes

- [ `x`] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by iizukak** [comment]
@tvlenin Thank you for PR!

**Comment by iizukak** [comment]
@ruimashita @nlpng Please review.

**Comment by ruimashita** [comment]
@tvlenin 
Thanks for this PR 👍👍👍  
When I try to change `space to depth`  to `maxpool` in LMNetV1 network,  the error comes in converting.

the error:
```
src/network.cpp: In member function ‘bool Network::run(float*, float*)’:
src/network.cpp:375:78: error: no matching function for call to ‘func_MaxPool(uint8_t*&, float*&, max_pooling_parameters&, int, int, int)’
 func_MaxPool(conv2_conv2d_Conv2D, pool2_MaxPool, MaxPool_struct, 127, 127, 64);
                                                                              ^
In file included from src/network.cpp:32:0:
./include/func/max_pool.h:50:6: note: candidate: void func_MaxPool(T_FLOAT*, T_FLOAT*, max_pooling_parameters, T_UINT, T_UINT, T_UINT)
 void func_MaxPool(T_FLOAT input[], T_FLOAT output[], struct max_pooling_parameters mpp, T_UINT out_height, T_UINT out_width, T_UINT out_depth);
      ^
./include/func/max_pool.h:50:6: note:   no known conversion for argument 1 from ‘uint8_t* {aka unsigned char*}’ to ‘T_FLOAT* {aka float*}’
In file included from src/network.cpp:32:0:
./include/func/max_pool.h:52:6: note: candidate: void func_MaxPool(T_INT*, T_INT*, max_pooling_parameters, T_UINT, T_UINT, T_UINT)
 void func_MaxPool(T_INT input[], T_INT output[], struct max_pooling_parameters mpp, T_UINT out_height, T_UINT out_width, T_UINT out_depth);
      ^
./include/func/max_pool.h:52:6: note:   no known conversion for argument 1 from ‘uint8_t* {aka unsigned char*}’ to ‘T_INT* {aka int*}’
```

Could you fix this error ?
Maybe I think, The reasones of this error  are
* Maxpool is `preserve_quantization` operation, But unfortunately the `MaxPool` class return preserve_quantization is `False` wrongly.
https://github.com/blue-oil/blueoil/blob/3711b3830ab00b30cdd8b328ca6452417d55b29f/dlk/python/dlk/core/operators.py#L1687

* Quantized data represent by `QUANTIZED_NOT_PACKED` (uint8), but unfortunately definitionof  input and output of maxpooling is `T_INT`(int) wrongly.
 https://github.com/blue-oil/blueoil/blob/3711b3830ab00b30cdd8b328ca6452417d55b29f/dlk/python/dlk/templates/src/func/max_pool.cpp#L117 https://github.com/blue-oil/blueoil/blob/3711b3830ab00b30cdd8b328ca6452417d55b29f/dlk/python/dlk/templates/include/func/max_pool.h#L52


---

train and convert command:
```
$ ./blueoil.sh train example/cls.yml ./saved/cls_maxpool
$ ./blueoil.sh convert example/cls.yml saved/cls_maxpool/cls_20190417165113/`
```

source code changes from this PR: https://github.com/ruimashita/blueoil/commit/e503d02fc591efed8ed0c3596662a69b126d9508

**Comment by tvlenin** [comment]
@ruimashita 
Thanks, I did what you recommend me and now is working, can you check again, please.

**Comment by ruimashita** [comment]
@tvlenin Thanks a lot 😄 🙇 😄 I will verify ✌️

**Comment by ruimashita** [comment]
I verified that
1. changed space to depth to maxpool in LMNetV1 network
2. train and convert
3. run `lm_fpga.elf ` on de10-nano


```root@DE10_NANO:~/cls_maxpool/cls_20190417165113/export/save.ckpt-27/128x128/output# ./models/lib/lm_fpga.elf ../inference_test_data/000_images_placeholder\:0.npy ../inference_test_data/133_output\:0.npy
-------------------------------------------------------------
Comparison: Default network test  succeeded!!!
-------------------------------------------------------------
AveragePool,5697,  sum:5.697ms
BatchNorm,4583,9313,  sum:13.896ms
Convolution,172537,86067,  sum:258.604ms
MaxPooling,210057,181467,450236,  sum:841.76ms
Packing input for kn2row,2176,4216,8362,4166,8095,  sum:27.015ms
QConv2D kn2row tiling,26893,52362,27189,37317,4567,  sum:148.328ms
QTZ_linear_mid_tread_half,5394,  sum:5.394ms
QuantizedConv2D,22318,  sum:22.318ms
QuantizedConv2D_ApplyScalingFactor,12610,  sum:12.61ms
ReLu,10835,  sum:10.835ms
SoftMax,122,  sum:0.122ms
TotalInitTime,98093,  sum:98.093ms
TotalRunTime,1.42799e+06,  sum:1427.99ms
kn2row,171319,  sum:171.319ms
kn2row-1x1,86055,  sum:86.055ms
kn2row-buf,3,  sum:0.003ms
matrix_multiplication,118291,86038,  sum:204.329ms
matrix_shift_add_f1,3569,  sum:3.569ms
matrix_shift_add_f2,49422,  sum:49.422ms
matrix_transpose (row_major),24,  sum:0.024ms
pack_input_to_qwords,2158,4201,8342,4151,8076,  sum:26.928ms
```

branch: https://github.com/ruimashita/blueoil/commit/dfbed67bd25010ed1e5352105d4b1e44512398b6

**Comment by ruimashita** [comment]
ok to test

File: dlk/python/dlk/templates/src/func/max_pool.cpp [hunk 1]
```diff
@@ -25,34 +26,42 @@ void max_pooling(
...
-            if(e > output[index_current_maximum_value])
-              output[index_current_maximum_value] = e;
+  assert (p.kernel_depth == 1 && "kernel depth 1 is not supported.");
+  assert (p.input_depth == p.kernel_depth * p.output_channels && \
+          "input_depth must equal kernel_depth * output_channels.");
+
+  int idx_out = 0;
+  const T_FLOAT num_k_elems = p.kernel_height * p.kernel_width * p.kernel_depth;
+
+  std::memset(output, 0.0f, p.output_channels * p.output_height * p.output_width * sizeof(TYPE));
```
Comment (tsawada) [inline dlk/python/dlk/templates/src/func/max_pool.cpp]:
```suggestion
  std::memset(output, 0, p.output_channels * p.output_height * p.output_width * sizeof(output[0]));
```

**Comment by tvlenin** [comment]
@tsawada @ruimashita 
Can you check again, I resolve the change request.
Thank you

**Pull request merged**